src/upgrades/mod.rs

```diff
@@ -159,13 +166,109 @@ pub trait UpgradeStore {
 
 /// Should be called when Krill starts, before the KrillServer is initiated
 pub fn pre_start_upgrade(config: Arc<Config>) -> Result<(), UpgradeError> {
-    upgrade_data_to_0_9_0(config)
+    upgrade_data_to_0_9_0(config.clone())?;
+
+    #[cfg(feature = "hsm")]
+    record_preexisting_openssl_keys_in_signer_mapper(config.clone())?;
+
+    Ok(())
+}
+
+/// Prior to Krill having HSM support there was no signer mapper as it wasn't needed, keys were just created by OpenSSL
+/// and stored in files on disk in KEYS_DIR named by the string form of their Krill KeyIdentifier. If Krill had created
+/// such keys and then the operator upgrades to a version of Krill with HSM support, the keys will become unusable
+/// because Krill will not be able to find a mapping from KeyIdentifier to signer as the mappings for the keys were
+/// never created. So we detect the case that the signer store SIGNERS_DIR directory has not yet been created, i.e. no
+/// signers have been registered and no key mappings have been recorded, and then walk KEYS_DIR adding the keys one by
+/// one to the mapping in the signer store.
+#[cfg(feature = "hsm")]
+fn record_preexisting_openssl_keys_in_signer_mapper(config: Arc<Config>) -> Result<(), UpgradeError> {
+    if !config.data_dir.join(SIGNERS_DIR).exists() {
+        let mut num_recorded_keys = 0;
+        let keys_dir = config.data_dir.join(KEYS_DIR);
+
+        info!(
+            "Scanning for not yet mapped OpenSSL signer keys in {} to record in the signer store",
+            keys_dir.to_string_lossy()
+        );
+
+        // For every file (key) in the legacy OpenSSL signer keys directory
+        if let Ok(dir_iter) = keys_dir.read_dir() {
+            let mut krill_signer: Option<KrillSigner> = None;
+            let mut signer_handle: Option<Handle> = None;
+
+            for entry in dir_iter {
+                let entry = entry.map_err(|err| {
+                    UpgradeError::IoError(KrillIoError::new(
+                        format!(
+                            "I/O error while looking for signer keys to register in: {}",
+                            keys_dir.to_string_lossy()
+                        ),
+                        err,
+                    ))
+                })?;
+
+                if entry.path().is_file() {
+                    // Is it a key identifier?
+                    if let Ok(key_id) = KeyIdentifier::from_str(&entry.file_name().to_string_lossy()) {
+                        if krill_signer.is_none() {
```

Comment:
The reason for creating it inside the loop is now no longer valid. It was created inside the loop because the creation of the KrillSigner caused Krill to immediately attempt to load the PKCS#11 library. I felt that this could be confusing if it happened during an upgrade step that turned out to be a no-op, why would it then have done anything concerning configured signers? However the PKCS#11 code has since been modified and no longer loads the library until first use. So I'll revisit this code in light of that.